pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -446,52 +453,128 @@ private String getOriginalPrincipal(String originalAuthData, String originalAuth
         return originalPrincipal;
     }
 
+    // complete the connect and sent newConnected command
+    private void completeConnect(int clientProtoVersion, String clientVersion) {
+        ctx.writeAndFlush(Commands.newConnected(clientProtoVersion));
+        state = State.Connected;
+        remoteEndpointProtocolVersion = clientProtoVersion;
+        if (isNotBlank(clientVersion) && !clientVersion.contains(" ") /* ignore default version: pulsar client */) {
+            this.clientVersion = clientVersion.intern();
+        }
+    }
+
+    // According to auth result, send newConnected or newAuthChallenge command.
+    private void doingAuthentication(AuthData clientData,
+                                     int clientProtocolVersion,
+                                     String clientVersion) throws Exception {
+        AuthData brokerData = authState.authenticate(clientData);
```

Comment:
This feels a little weird, where you're returning an object, from the authState, then completely ignoring it if auth is complete. Maybe instead of returning AuthData, authenticate could return AuthResult.

```
class AuthResult {
   Status getStatus(); // can return COMPLETE/IN_PROGRESS/ERROR
   AuthData getData();
}
```

Reply:
Stats(complete/inprogress/error) in kept in AuthState. AuthState.authenticate() is not only return authData, it is also change the state of AuthState during AuthState.authenticate(). So at last the status is kept in AuthState.